### PR TITLE
983: Upgrade dependencies

### PR DESCRIPTION
**Pull request by steffenkleinle**
### Short description

Upgrade most dependencies.

### Proposed changes

- Upgrade most dependencies
- Fix jest setup and align jest.config.ts to integreat
- Extract and fix `getGalleryPermission()`
- Fix a few wrong and tests asserting on styles
- Fix some types
- Uncomment tests with `ByRole` queries, should be fixed in react-native upgrade (#1010)

Separate PR for an upgrade of prettier: #1014.

Created separate tickets for upgrading
- react-native including react-native-reanimated and react-native-gesture-handler (#1010)
- victory-native (#1011)

Tickets already existing for upgrades/migrations:
- Remove patch and upgrade react-native-audio-record-player: #890
- Replace react-native-camera: #873

Ticket for unmaintained dependencies:
https://github.com/digitalfabrik/lunes-app/issues/1015.

### Side effects

Hopefully none.

### Resolved issues

Fixes: #983.

---

### CODE REVIEW

File: src/components/VocabularyListItem.tsx [hunk 1]
```diff
@@ -9,6 +9,7 @@ import FavoriteButton from './FavoriteButton'
 import ListItem from './ListItem'
 import { ContentTextLight } from './text/Content'
 
+// TODO check width (md)
```
Comment (ztefanie) [inline src/components/VocabularyListItem.tsx:12]: have you checked this?

File: src/routes/write-exercise/services/AbstractWriteExerciseService.ts [hunk 2]
```diff
@@ -8,13 +8,13 @@ import { RoutesParams, VocabularyItemResult } from '../../../navigation/Navigati
...
 export default abstract class AbstractWriteExerciseService {
-  // eslint-disable-next-line no-useless-constructor
+  // eslint-disable-next-line no-useless-constructor,no-empty-function
   constructor(
     public route: RouteProp<RoutesParams, 'WriteExercise'>,
     public navigation: StackNavigationProp<RoutesParams, 'WriteExercise'>,
     public setCurrentIndex: React.Dispatch<React.SetStateAction<number>>,
     public setIsAnswerSubmitted: React.Dispatch<React.SetStateAction<boolean>>,
-    public setVocabularyItemWithResults: React.Dispatch<React.SetStateAction<VocabularyItemResult[]>>
+    public setVocabularyItemWithResults: React.Dispatch<React.SetStateAction<VocabularyItemResult[]>> // eslint-disable-next-line no-empty-function
```
Comment (ztefanie) [inline src/routes/write-exercise/services/AbstractWriteExerciseService.ts:17]: I this this eslint comment should be removed?

**Review by ztefanie — APPROVED** [review]
Tested on Android, works as expected :rocket: 
Thanks for updating :pray:

File: src/components/VocabularyListItem.tsx (hunk 1 above)
Comment (steffenkleinle) [reply src/components/VocabularyListItem.tsx:12]: Thanks for the reminder, not yet.

File: src/routes/write-exercise/services/AbstractWriteExerciseService.ts (hunk 2 above)
Comment (steffenkleinle) [reply src/routes/write-exercise/services/AbstractWriteExerciseService.ts:17]: Actually no, its needed bc of the empty constructor.

Comment (ztefanie) [reply src/routes/write-exercise/services/AbstractWriteExerciseService.ts:17]: Then you can delete it in line 11, currently it is duplicated disabled.

File: src/routes/write-exercise/services/AbstractWriteExerciseService.ts [hunk 3]
```diff
@@ -8,13 +8,13 @@ import { RoutesParams, VocabularyItemResult } from '../../../navigation/Navigati
 import { moveToEnd, shuffleArray } from '../../../services/helpers'
 
 export default abstract class AbstractWriteExerciseService {
-  // eslint-disable-next-line no-useless-constructor
+  // eslint-disable-next-line no-useless-constructor,no-empty-function
```
Comment (ztefanie) [inline src/routes/write-exercise/services/AbstractWriteExerciseService.ts:11]:
```suggestion
  // eslint-disable-next-line no-useless-constructor
```

**Review by ztefanie — APPROVED** [review]

**Review by LeandraH — APPROVED** [review]
Very cool, thank you! Tested on real iOS
There were a few issues that I had to solve to get it running on iOS (I'll push them in a minute):
- Update cocoapods (https://github.com/getsentry/sentry-react-native/issues/3547)
- Update permissions handling (https://github.com/zoontek/react-native-permissions/releases/tag/3.9.0)

Would appreciate someone else checking on iOS @ztefanie @f1sh1918

**Comment by ztefanie** [comment]
Does not work for me on mac,
Screenshot 2024-04-08 at 09 58 11

I deleted the pods folder, cleared the pod cache and reinstalled everything.

**Comment by ztefanie** [comment]
Could be due to outdated OS (and therefore xcode version) of my mac.

File: jest.config.ts [hunk 4]
```diff
@@ -1,28 +1,20 @@
...
+  roots: ['src'],
   preset: 'react-native',
   verbose: true,
-  /* Always explicitly mock modules. Also automocking seems to be broken right now:
-        https://github.com/facebook/jest/issues/6127 */
   automock: false,
+  moduleDirectories: ['node_modules'],
   moduleNameMapper: {
-    '.+\\.(styl|sass|scss|png|jpg|ttf|woff|woff2)$': 'jest-transform-stub',
-    '\\.(css|less)$': 'identity-obj-proxy',
```
[the remark this answers is not in the document]
Comment (steffenkleinle) [reply jest.config.ts:11]: Good catch!

File: jest.config.ts [hunk 5]
```diff
@@ -1,28 +1,20 @@
...
-    '\\.svg': '<rootDir>/__mocks__/svgrMock.js',
-    '^axios$': 'axios/dist/axios.js',
+    '.+\\.(css|styl|less|sass|scss|png|jpg|ttf|woff|woff2)$': 'jest-transform-stub',
```
[the remark this answers is not in the document]
Comment (steffenkleinle) [reply jest.config.ts:10]: This is just copied from the `jest-transform-stub`/`integreat-app` so I'd just keep it this way as it doesn't hurt.

File: src/services/AsyncStorage.ts [hunk 6]
```diff
@@ -215,9 +218,11 @@ export const deleteUserVocabularyItem = async (userVocabularyItem: VocabularyIte
...
   )
   const images = userVocabularyItem.images.map(image => image.image)
-  images.map(async image => {
-    await unlink(image)
-  })
+  await Promise.all(
+    images.map(async image => {
+      await unlink(image)
+    })
+  )
```
[the remark this answers is not in the document]
Comment (steffenkleinle) [reply src/services/AsyncStorage.ts:225]: Yes, as `Promise.all` returns a promise as well. If that promise is not awaited, there is no guarantee that the other promises are already fulfilled.

See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all:
> The Promise.all() static method takes an iterable of promises as input and returns a single [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise). This returned promise fulfills when all of the input's promises fulfill (including when an empty iterable is passed), with an array of the fulfillment values. It rejects when any of the input's promises rejects, with this first rejection reason.

File: src/components/VocabularyListItem.tsx (hunk 1 above)
Comment (steffenkleinle) [reply src/components/VocabularyListItem.tsx]: Checked, works fine.

**Comment by steffenkleinle** [comment]
@ztefanie @LeandraH should I wait with merging until you have a new mac and can test this?

**Comment by ztefanie** [comment]
> @ztefanie @LeandraH should I wait with merging until you have a new mac and can test this?

Maybe @f1sh1918 can test it?

**Comment by f1sh1918** [comment]
leandras approve should be sufficient? @steffenkleinle

**Comment by LeandraH** [comment]
Since I now wrote part of the code, I feel like another person should test the part that I wrote. But that might be paranoia on my part :)

**Comment by steffenkleinle** [comment]
I am fine with both, will be tested by Martin before a production release again anyway right? Also @LeandraH I think you just made sure that dependencies are correctly picked up for iOS, you didn't change any user-visible code. So I think we could merge this from my perspective.

**Comment by LeandraH** [comment]
Works for me

**Pull request merged**